### PR TITLE
Model Updates

### PR DESCRIPTION
**Pull request by trahxam**
Fix some small bugs:
- kv mask sorting bug
- bug where batch size of 1 with flash-varlen would cause error
- nan loss with masks with zero constituents
- lrsm dataset bug
- simplistic multi task learning

### CODE REVIEW

File: src/hepattn/models/wrapper.py [hunk 1]
```diff
@@ -28,32 +29,39 @@ def __init__(
...
         return self.model.predict(outputs)
 
-    def log_losses(self, losses, stage):
-        total_loss = 0
+    def aggregate_losses(self, losses: dict[str, Tensor], stage: str | None = None) -> Tensor:
+        device = next(self.model.parameters()).device
+        total_loss = torch.tensor(0.0, device=device)
 
         for layer_name, layer_losses in losses.items():
             layer_loss = 0
```
Comment (samvanstroud) [inline src/hepattn/models/wrapper.py]: Was this a necessary change? probably best to be consistent with the `layer_loss` (even if it's just for logging atm).

File: src/hepattn/models/loss.py [hunk 2]
```diff
@@ -194,36 +215,53 @@ def mask_focal_cost(pred_logits, targets, gamma=2.0, input_pad_mask=None, sample
...
 
 
-def mask_bce_loss(pred_logits, targets, object_valid_mask=None, input_pad_mask=None, sample_weight=None):
-    """Compute the binary cross-entropy loss for binary masks.
+def mask_bce_loss(
+    pred_logits: torch.Tensor,
+    targets: torch.Tensor,
+    object_valid_mask: torch.Tensor | None = None,  # [batch_size, num_objects]
+    input_pad_mask: torch.Tensor | None = None,  # [batch_size, num_constituents]
+    sample_weight: torch.Tensor | None = None,  # broadcastable to [batch_size, num_objects, num_constituents]
```
Comment (samvanstroud) [inline src/hepattn/models/loss.py]: maybe move/harmoinse these to the docstring? will be more useful there if we auto generate docs

File: src/hepattn/models/loss.py [hunk 3]
```diff
@@ -194,36 +215,53 @@ def mask_focal_cost(pred_logits, targets, gamma=2.0, input_pad_mask=None, sample
...
+    inp_mask = input_pad_mask.unsqueeze(1).to(loss.dtype)
+    loss = loss * obj_mask * inp_mask
+
+    valid_inputs = inp_mask.sum(-1).clamp_min(1.0)
+    per_obj = loss.sum(-1) / valid_inputs
+
+    valid_objs = object_valid_mask.sum(-1).clamp_min(1)
+    per_batch = (per_obj * object_valid_mask).sum(-1) / valid_objs
+
+    return per_batch.mean()
```
Comment (samvanstroud) [inline src/hepattn/models/loss.py]: is this change just to avoid nans for zero constituents, and otherwise leaves the loss unchanged? since I think multiplying by the `object_valid_mask` is less performant than indexing by it early, maybe we can just check the zero constituent case manually?

Having said that this impl is definitely cleaner, so if the perf impact is not large we could go with this. I know focal at least is pretty heavy on memory though,,

File: src/hepattn/models/wrapper.py [hunk 4]
```diff
@@ -28,32 +29,39 @@ def __init__(
...
+    def forward(self, inputs: dict[str, Tensor]) -> dict[str, Tensor]:
         return self.model(inputs)
 
-    def predict(self, outputs):
+    def predict(self, outputs: dict[str, Tensor]) -> dict[str, Tensor]:
         return self.model.predict(outputs)
 
-    def log_losses(self, losses, stage):
-        total_loss = 0
+    def aggregate_losses(self, losses: dict[str, Tensor], stage: str | None = None) -> Tensor:
```
Comment (samvanstroud) [inline src/hepattn/models/wrapper.py]: I think losses is a nested dict no?

File: src/hepattn/models/loss.py (hunk 2 above)
Comment (samvanstroud) [inline src/hepattn/models/loss.py]: I think best to leave this in the docstring, we might add autodocs at some point and don't want this going out of sync

File: src/hepattn/models/maskformer.py [hunk 5]
```diff
@@ -88,8 +88,9 @@ def forward(self, inputs: dict[str, Tensor]) -> tuple[dict[str, Tensor], dict[st
         x["key_embed"] = torch.concatenate([x[input_name + "_embed"] for input_name in self.input_names], dim=-2)
         x["key_valid"] = torch.concatenate([x[input_name + "_valid"] for input_name in self.input_names], dim=-1)
 
-        # if all key_valid are true, then we can just set it to None
-        if batch_size == 1 and x["key_valid"].all():
+        # If all key_valid are true, then we can just set it to None, however,
+        # if we are using flash-varlen, we have to always provide a kv_mask argument
+        if batch_size == 1 and x["key_valid"].all() and self.encoder.attn_type != "flash_varlen":
```
[the remark this answers is not in the document]
Comment (trahxam) [reply src/hepattn/models/maskformer.py]: The issue is that if you have a batch size of >1, you can still end up with a batch size of 1, I think? e.g. if you have 101 samples and a batch size of 50.

File: src/hepattn/models/wrapper.py (hunk 1 above)
Comment (trahxam) [reply src/hepattn/models/wrapper.py]: duplicate

File: src/hepattn/models/loss.py [hunk 6]
```diff
@@ -128,42 +128,63 @@ def mask_iou_cost(pred_logits, targets, input_pad_mask=None, eps=1e-6):
...
+    inp_mask = input_pad_mask.unsqueeze(1).to(loss.dtype)
+    loss = loss * obj_mask * inp_mask
+
+    valid_inputs = inp_mask.sum(-1).clamp_min(1.0)
+    per_obj = loss.sum(-1) / valid_inputs
+
+    valid_objs = object_valid_mask.sum(-1).clamp_min(1)
+    per_batch = (per_obj * object_valid_mask.to(per_obj.dtype)).sum(-1) / valid_objs
+
+    return per_batch.mean()
```
[the remark this answers is not in the document]
Comment (samvanstroud) [reply src/hepattn/models/loss.py]: Going to keep this open until you have time to implement this, since I think this could actually have a bit of a large impact on @pippaduckett1's work on trackml. No rush though!
